packages/app/src/components/Randomizer/Randomizer.js

```diff
@@ -1,58 +1,111 @@
+import React, { useState, useEffect } from 'react'
 import { Button } from 'react-bootstrap'
 import { ExpandMore, ExpandLess } from '@material-ui/icons'
+import PropTypes from 'prop-types'
 import Filter from '../Filter/Filter/Filter'
 import RestaurantEntry from '../RestaurantEntry/RestaurantEntry'
 import soundService from '../../services/sound'
 import restaurantService from '../../services/restaurant'
 import './Randomizer.css'
 
 
-const Randomizer = () => {
-  const maxNumberOfRotations = 28
-  const minTimeBetweenRotations = 15 // in milliseconds
-  const [restaurant, setRestaurant] = useState({ name: 'Press the button' })
-  const [disableButton, setDisableButton] = useState(false)
+const Randomizer = ({ maxNumberOfRotations = 35, minTimeBetweenRotations = 25, maxTimeBetweenRotations = 1000, resultWaitTime = 1250 }) => {
+  const [restaurant, setRestaurant] = useState()
+  const [isRolling, setRolling] = useState(false)
   const [filterType, setFilterType] = useState('some')
   const [filterCategories, setFilterCategories] = useState([])
   const [distance, setDistance] = useState('')
   const [showFilter, setShowFilter] = useState(false)
 
-  const sleep = async (duration) => {
-    return new Promise(r => setTimeout(r, duration))
+  const [timeoutHandle, setTimeoutHandle] = useState()
+  const [nope, setNope] = useState()
+
+  useEffect(() => {
+    return () => clearTimeout(timeoutHandle)
+  }, [timeoutHandle])
+
+  const shuffle = (array) => {
+    for (let i = array.length - 1; i > 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [array[i], array[j]] = [array[j], array[i]]
+    }
+    return array
   }
 
-  const roll = async (event) => {
+  const easingFunc = (max, min, t) => {
+    //return (1 - t) * min + t * max // Linear interpolation
+    return (max - min) * Math.pow(t, 4) + min
+  }
+
+  const rollNext = (iteration, restaurants) => {
+    const count = restaurants.length
+    const nextRestaurant = restaurants[iteration % count]
+    setRestaurant(nextRestaurant)
+
+    const iterationsRemaining = maxNumberOfRotations - iteration
+    if (iterationsRemaining === 0) {
+      if (Math.random() < 0.025) {
+        setNope(true)
+        setTimeoutHandle(setTimeout(() => rollNext(Math.ceil(maxNumberOfRotations / 2), shuffle(restaurants)), minTimeBetweenRotations))
+        return
+      } else {
+        soundService.playFanfare()
+        setRolling(false)
+        setTimeoutHandle()
+        return
+      }
+    }
+
+    if (iterationsRemaining < Math.ceil(maxNumberOfRotations / 2) - 5) {
+      setNope(false)
+    }
+
+    soundService.playBeep()
+
+    const t = iteration / maxNumberOfRotations
+    let interpolatedTime = easingFunc(maxTimeBetweenRotations, minTimeBetweenRotations, t)
+    if (iterationsRemaining === 1) {
+      interpolatedTime += resultWaitTime
+    }
+
+    setTimeoutHandle(setTimeout(() => rollNext(iteration + 1, restaurants), interpolatedTime))
+  }
```

Comment:
Is there any way to make this shorter / prettier?

Reply:
Yeah, it's ugly. Re-wrote the whole damn thing.